sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ShouldImplementExportedInterfaces.cs

```diff
@@ -0,0 +1,142 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class ShouldImplementExportedInterfaces : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S4159";
+        private const string MessageFormat = "Implement '{0}' on '{1}' or remove this Export attribute.";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c =>
+                {
+                    var attributeSyntax = (AttributeSyntax)c.Node;
+
+                    var attributeCtorSymbol = c.SemanticModel.GetSymbolInfo(attributeSyntax.Name).Symbol as IMethodSymbol;
+                    if (attributeCtorSymbol == null ||
+                        !attributeCtorSymbol.ContainingType.Is(KnownType.System_ComponentModel_Composition_ExportAttribute))
+                    {
+                        return;
+                    }
+
+                    var exportedType = GetExportedTypeSyntax(attributeSyntax, c.SemanticModel);
+                    var attributeTargetType = GetAttributeTargetSymbol(attributeSyntax, c.SemanticModel);
+
+                    if (exportedType != null &&
+                        attributeTargetType != null &&
+                        !attributeTargetType.DerivesOrImplements(exportedType))
+                    {
+                        c.ReportDiagnosticWhenActive(
+                            Diagnostic.Create(rule, attributeSyntax.GetLocation(), exportedType.Name, attributeTargetType.Name));
+                    }
+                },
+                SyntaxKind.Attribute);
+        }
+
+        private ITypeSymbol GetAttributeTargetSymbol(AttributeSyntax attributeSyntax, SemanticModel semanticModel)
+        {
+            // Parent is AttributeListSyntax, we handle only class attributes
+            var attributeTarget = attributeSyntax.Parent?.Parent as ClassDeclarationSyntax;

```

Comment:
Not sure this will cover parenthesis. I would use first ancestor of type `ClassDeclarationSyntax`.

Reply:
This will catch method attributes too, because it will go up to the class declaration. Also I don't think you could have class declaration with parenthesis.